Preserve naming for inputs/outputs with observer insertion

nishantpdce opened the pull request:
Summary:
- Quantizer observer node output is hooked up to following node
which mutates the naming for input/output. This is not desired and
required because observer op can be a sync node

  - Quantizer is aimed for quantizing tensors so we should insert observer
op for Values that are type tensor

Differential Revision: D14715916

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,91 @@
+  Node* quant = qdq.first;
+  Node* dequant = qdq.second;
 
+  // Add quant-dequant nodes to the graph at appropriate insertion points
```
'appropriate insertion points' -- can you expand on this? The comment doesn't help me understand what an appropriate point is for quant-dequant nodes on inputs.

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -1,14 +1,91 @@
+  quant->insertBefore(dequant);
+  n->replaceInputWith(v, dequant->outputs()[0]);
+
+  // Attach inputs to quant and dequant nodes
```
Again, this comment explains what the code does (which I can read from the code). Can it be change to explain _why_?

zdevito commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -83,12 +156,95 @@ void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+          }
+        }
+      } else {
+        // This iteration is similar to external inputs to block.
```
Comment just explains code and does not explain the _goal_ of the code.

zdevito reviewed (commented):
Hi -- see the inline comments. I am still having trouble understand what is being accomplished by reading the comments.

nishantpdce committed:
Preserve naming for inputs/outputs with observer insertion (#18713)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/18713

  - Quantizer observer node output is hooked up to following node
which mutates the naming for input/output. This is not desired and
required because observer op can be a sync node

  - Quantizer is aimed for quantizing tensors so we should insert observer
op for Values that are type tensor

Reviewed By: zafartahirov

Differential Revision: D14715916

fbshipit-source-id: f5e90e60e65e739057d087c718e2faef709018d8

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -125,10 +125,6 @@ static void addObserverFor(Value* v, Node* original_observer_node) {
 
-  // Replace the uses of v with observedValue. We need to do it *before* we add
-  // the inputs - otherwise we would replace the newly added inputs as well.
-  v->replaceAllUsesWith(observedValue);
```
Could you please expand on why this is needed to be removed?
